From 49edf344eb1734f3b51671f53ff0e69a3962cfa6 Mon Sep 17 00:00:00 2001
From: Philippe Gerum <rpm@xenomai.org>
Date: Mon, 17 Jul 2017 09:35:14 +0200
Subject: [PATCH] hrtimer: dovetail: announce clock set event

The co-kernel needs to be told when the real-time epoch changes, in
order to adjust its own timers.
---
 include/linux/dovetail.h | 4 ++++
 kernel/dovetail.c        | 4 ++++
 kernel/time/hrtimer.c    | 1 +
 3 files changed, 9 insertions(+)

diff --git a/include/linux/dovetail.h b/include/linux/dovetail.h
index 2c06473c68b4..55010e66c3cd 100644
--- a/include/linux/dovetail.h
+++ b/include/linux/dovetail.h
@@ -52,6 +52,8 @@ static inline void oob_trap_notify(unsigned int trapnr,
 void inband_event_notify(enum inband_event_type,
 			 void *data);
 
+void inband_clock_was_set(void);
+
 static inline void inband_signal_notify(struct task_struct *p)
 {
 	if (test_ti_local_flags(task_thread_info(p), _TLF_DOVETAIL))
@@ -220,6 +222,8 @@ static inline int inband_switch_tail(void)
 #define unprotect_inband_mm(__flags)	\
 	do { (void)(__flags); } while (0)
 
+static inline void inband_clock_was_set(void) { }
+
 #endif	/* !CONFIG_DOVETAIL */
 
 static inline bool dovetailing(void)
diff --git a/kernel/dovetail.c b/kernel/dovetail.c
index 6da6429bdf5e..05ba868cda1e 100644
--- a/kernel/dovetail.c
+++ b/kernel/dovetail.c
@@ -319,6 +319,10 @@ int inband_switch_tail(void)
 	return !inband;
 }
 
+void __weak inband_clock_was_set(void)
+{
+}
+
 int dovetail_start(void)
 {
 	check_inband_stage();
diff --git a/kernel/time/hrtimer.c b/kernel/time/hrtimer.c
index 5ee77f1a8a92..376486815c62 100644
--- a/kernel/time/hrtimer.c
+++ b/kernel/time/hrtimer.c
@@ -851,6 +851,7 @@ void clock_was_set(void)
 	on_each_cpu(retrigger_next_event, NULL, 1);
 #endif
 	timerfd_clock_was_set();
+	inband_clock_was_set();
 }
 
 /*
-- 
2.16.4

